File: src/main/java/com/pedigree/ui/FamilyDialog.java
package com.pedigree.ui;

import com.pedigree.model.Family;
import com.pedigree.model.Individual;
import com.pedigree.model.Gender;
import com.pedigree.storage.ProjectRepository;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Priority;
import javafx.scene.layout.VBox;
import javafx.scene.layout.ColumnConstraints;
import javafx.scene.layout.Region;
import javafx.stage.Modality;
import javafx.stage.Stage;

import java.util.Optional;

public class FamilyDialog {

    public Optional<Family> showCreate(ProjectRepository.ProjectData data) {
        if (data == null) return Optional.empty();
        Stage stage = new Stage();
        stage.initModality(Modality.APPLICATION_MODAL);
        stage.setTitle("New Family");

        GridPane grid = buildForm(null, data);
        VBox root = new VBox(grid);
        root.setPadding(new Insets(8));
        Scene scene = new Scene(root, 640, 520);
        stage.setScene(scene);
        stage.setResizable(true);
        stage.setMinWidth(580);
        stage.setMinHeight(460);

        final Family[] result = new Family[1];
        Button btnOk = new Button("Create");
        Button btnCancel = new Button("Cancel");
        HBox buttons = new HBox(10, btnOk, btnCancel);
        buttons.setAlignment(Pos.CENTER_RIGHT);
        GridPane.setColumnSpan(buttons, 2);
        grid.add(buttons, 0, 5);

        btnOk.setDefaultButton(true);
        btnCancel.setCancelButton(true);

        @SuppressWarnings("unchecked")
        ComboBox<Individual> cbA = (ComboBox<Individual>) grid.lookup("#husband");
        @SuppressWarnings("unchecked")
        ComboBox<Individual> cbB = (ComboBox<Individual>) grid.lookup("#wife");
        TextField tfMarriageDate = (TextField) grid.lookup("#marriageDate");
        Button btnMarriageDate = (Button) grid.lookup("#marriageDatePicker");
        TextField tfMarriagePlace = (TextField) grid.lookup("#marriagePlace");
        @SuppressWarnings("unchecked")
        ListView<Individual> lvChildrenSelected = (ListView<Individual>) grid.lookup("#childrenSelected");

        DatePhraseDialog dateDlg = new DatePhraseDialog();
        btnMarriageDate.setOnAction(ev -> dateDlg.show(tfMarriageDate.getText()).ifPresent(tfMarriageDate::setText));

        btnOk.setOnAction(e -> {
            Family fam = new Family();
            Individual a = cbA.getValue();
            Individual b = cbB.getValue();
            fam.setHusbandId(a != null ? a.getId() : null);
            fam.setWifeId(b != null ? b.getId() : null);
            fam.setMarriageDate(tfMarriageDate.getText() != null ? tfMarriageDate.getText().trim() : null);
            fam.setMarriagePlace(tfMarriagePlace.getText() != null ? tfMarriagePlace.getText().trim() : null);
            lvChildrenSelected.getItems()
                    .forEach(child -> fam.getChildrenIds().add(child.getId()));
            result[0] = fam;
            com.pedigree.util.DirtyFlag.setModified();
            stage.close();
        });
        btnCancel.setOnAction(e -> stage.close());

        stage.showAndWait();
        return Optional.ofNullable(result[0]);
    }

    public boolean showEdit(Family fam, ProjectRepository.ProjectData data) {
        if (fam == null || data == null) return false;
        Stage stage = new Stage();
        stage.initModality(Modality.APPLICATION_MODAL);
        stage.setTitle("Edit Family");

        GridPane grid = buildForm(fam, data);
        VBox root = new VBox(grid);
        root.setPadding(new Insets(8));
        Scene scene = new Scene(root, 640, 520);
        stage.setScene(scene);
        stage.setResizable(true);
        stage.setMinWidth(580);
        stage.setMinHeight(460);

        final boolean[] saved = new boolean[]{false};
        Button btnOk = new Button("Save");
        Button btnCancel = new Button("Cancel");
        HBox buttons = new HBox(10, btnOk, btnCancel);
        buttons.setAlignment(Pos.CENTER_RIGHT);
        GridPane.setColumnSpan(buttons, 2);
        grid.add(buttons, 0, 5);

        btnOk.setDefaultButton(true);
        btnCancel.setCancelButton(true);

        @SuppressWarnings("unchecked")
        ComboBox<Individual> cbA = (ComboBox<Individual>) grid.lookup("#husband");
        @SuppressWarnings("unchecked")
        ComboBox<Individual> cbB = (ComboBox<Individual>) grid.lookup("#wife");
        TextField tfMarriageDate = (TextField) grid.lookup("#marriageDate");
        Button btnMarriageDate = (Button) grid.lookup("#marriageDatePicker");
        TextField tfMarriagePlace = (TextField) grid.lookup("#marriagePlace");
        @SuppressWarnings("unchecked")
        ListView<Individual> lvChildrenSelected = (ListView<Individual>) grid.lookup("#childrenSelected");

        DatePhraseDialog dateDlg = new DatePhraseDialog();
        btnMarriageDate.setOnAction(ev -> dateDlg.show(tfMarriageDate.getText()).ifPresent(tfMarriageDate::setText));

        btnOk.setOnAction(e -> {
            Individual a = cbA.getValue();
            Individual b = cbB.getValue();
            fam.setHusbandId(a != null ? a.getId() : null);
            fam.setWifeId(b != null ? b.getId() : null);
            fam.setMarriageDate(tfMarriageDate.getText() != null ? tfMarriageDate.getText().trim() : null);
            fam.setMarriagePlace(tfMarriagePlace.getText() != null ? tfMarriagePlace.getText().trim() : null);
            fam.getChildrenIds().clear();
            lvChildrenSelected.getItems()
                    .forEach(child -> fam.getChildrenIds().add(child.getId()));
            saved[0] = true;
            com.pedigree.util.DirtyFlag.setModified();
            stage.close();
        });
        btnCancel.setOnAction(e -> stage.close());

        stage.showAndWait();
        return saved[0];
    }

    private GridPane buildForm(Family existing, ProjectRepository.ProjectData data) {
        GridPane grid = new GridPane();
        grid.setHgap(10);
        grid.setVgap(10);
        grid.setPadding(new Insets(12));
        // Column 0 = labels (no grow), Column 1 = controls/content (grow)
        ColumnConstraints col0 = new ColumnConstraints();
        col0.setHgrow(Priority.NEVER);
        col0.setFillWidth(false);
        ColumnConstraints col1 = new ColumnConstraints();
        col1.setHgrow(Priority.ALWAYS);
        grid.getColumnConstraints().addAll(col0, col1);

        Label lA = new Label("Husband:");
        lA.setMinWidth(Region.USE_PREF_SIZE);
        ComboBox<Individual> cbA = new ComboBox<>();
        cbA.setId("husband");
        // Only males can be selected as husband
                for (Individual i : data.individuals) {
                    if (i.getGender() == Gender.MALE) {
                        cbA.getItems().add(i);
                    }
                }
        cbA.setCellFactory(list -> new ListCell<>() {
            @Override protected void updateItem(Individual item, boolean empty) {
                super.updateItem(item, empty);
                setText(empty || item == null ? null : displayName(item));
            }
        });
        cbA.setButtonCell(new ListCell<>() {
            @Override protected void updateItem(Individual item, boolean empty) {
                super.updateItem(item, empty);
                setText(empty || item == null ? null : displayName(item));
            }
        });
        // Also set a converter to ensure value is always displayed even if not in items
        cbA.setConverter(new javafx.util.StringConverter<>() {
            @Override public String toString(Individual object) {
                return object == null ? "" : displayName(object);
            }
            @Override public Individual fromString(String string) { return null; }
        });

        Label lB = new Label("Wife:");
        lB.setMinWidth(Region.USE_PREF_SIZE);
        ComboBox<Individual> cbB = new ComboBox<>();
        cbB.setId("wife");
        // Only females can be selected as wife
                for (Individual i : data.individuals) {
                    if (i.getGender() == Gender.FEMALE) {
                        cbB.getItems().add(i);
                    }
                }
        cbB.setCellFactory(cbA.getCellFactory());
        cbB.setButtonCell(new ListCell<>() {
            @Override protected void updateItem(Individual item, boolean empty) {
                super.updateItem(item, empty);
                setText(empty || item == null ? null : displayName(item));
            }
        });
        cbB.setConverter(new javafx.util.StringConverter<>() {
            @Override public String toString(Individual object) {
                return object == null ? "" : displayName(object);
            }
            @Override public Individual fromString(String string) { return null; }
        });

        Label lMarriageDate = new Label("Marriage Date:");
        lMarriageDate.setMinWidth(Region.USE_PREF_SIZE);
        TextField tfMarriageDate = new TextField(existing != null ? nullToEmpty(existing.getMarriageDate()) : "");
        tfMarriageDate.setId("marriageDate");
        tfMarriageDate.setPrefWidth(300);
        Button btnMarriageDate = new Button("...");
        btnMarriageDate.setId("marriageDatePicker");
        HBox hbMarriageDate = new HBox(6, tfMarriageDate, btnMarriageDate);

        Label lMarriagePlace = new Label("Marriage Place:");
        lMarriagePlace.setMinWidth(Region.USE_PREF_SIZE);
        TextField tfMarriagePlace = new TextField(existing != null ? nullToEmpty(existing.getMarriagePlace()) : "");
        tfMarriagePlace.setId("marriagePlace");
        tfMarriagePlace.setPrefWidth(300);

        Label lChildren = new Label("Children:");
        lChildren.setMinWidth(Region.USE_PREF_SIZE);
        // Dual-list UI: Available individuals -> Selected children
        ListView<Individual> lvChildrenAvailable = new ListView<>();
        lvChildrenAvailable.setId("childrenAvailable");
        lvChildrenAvailable.getItems().addAll(data.individuals);
        lvChildrenAvailable.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);
        lvChildrenAvailable.setCellFactory(list -> new ListCell<>() {
            @Override protected void updateItem(Individual item, boolean empty) {
                super.updateItem(item, empty);
                setText(empty || item == null ? null : displayName(item));
            }
        });

        ListView<Individual> lvChildrenSelected = new ListView<>();
        lvChildrenSelected.setId("childrenSelected");
        lvChildrenSelected.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);
        lvChildrenSelected.setCellFactory(list -> new ListCell<>() {
            @Override protected void updateItem(Individual item, boolean empty) {
                super.updateItem(item, empty);
                setText(empty || item == null ? null : displayName(item));
            }
        });

        Button btnAddChild = new Button("Add ▶");
        Button btnRemoveChild = new Button("◀ Remove");
        btnAddChild.setOnAction(e -> {
            var selected = new java.util.ArrayList<>(lvChildrenAvailable.getSelectionModel().getSelectedItems());
            for (Individual ind : selected) {
                if (!lvChildrenSelected.getItems().contains(ind)) {
                    lvChildrenSelected.getItems().add(ind);
                }
            }
            lvChildrenAvailable.getSelectionModel().clearSelection();
        });
        btnRemoveChild.setOnAction(e -> {
            var selected = new java.util.ArrayList<>(lvChildrenSelected.getSelectionModel().getSelectedItems());
            lvChildrenSelected.getItems().removeAll(selected);
        });

        // Double-click convenience
        lvChildrenAvailable.setOnMouseClicked(e -> {
            if (e.getClickCount() == 2) btnAddChild.fire();
        });
        lvChildrenSelected.setOnMouseClicked(e -> {
            if (e.getClickCount() == 2) btnRemoveChild.fire();
        });

        // Initialize existing values and keep spouses out of children
        if (existing != null) {
            if (existing.getHusbandId() != null) {
                data.individuals.stream()
                        .filter(i -> i.getId().equals(existing.getHusbandId()))
                        .findFirst()
                        .ifPresent(h -> {
                            // Ensure the selected husband is present in the items even if filtered out (e.g., UNKNOWN gender)
                            if (!cbA.getItems().contains(h)) {
                                cbA.getItems().add(0, h);
                            }
                            cbA.setValue(h);
                        });
            }
            if (existing.getWifeId() != null) {
                data.individuals.stream()
                        .filter(i -> i.getId().equals(existing.getWifeId()))
                        .findFirst()
                        .ifPresent(w -> {
                            if (!cbB.getItems().contains(w)) {
                                cbB.getItems().add(0, w);
                            }
                            cbB.setValue(w);
                        });
            }
            for (String cid : existing.getChildrenIds()) {
                data.individuals.stream().filter(i -> i.getId().equals(cid)).findFirst().ifPresent(i -> {
                    if (!lvChildrenSelected.getItems().contains(i)) {
                        lvChildrenSelected.getItems().add(i);
                    }
                });
            }
        }

        // Ensure spouses are not listed as available children and are removed if selected
        Runnable refreshAvailable = () -> {
            lvChildrenAvailable.getItems().setAll(data.individuals);
            Individual a = cbA.getValue();
            Individual b = cbB.getValue();
            if (a != null) lvChildrenAvailable.getItems().remove(a);
            if (b != null) lvChildrenAvailable.getItems().remove(b);
            // Also make sure spouses are not in selected
            if (a != null) lvChildrenSelected.getItems().remove(a);
            if (b != null) lvChildrenSelected.getItems().remove(b);
            // And avoid duplicates between lists
            lvChildrenAvailable.getItems().removeAll(lvChildrenSelected.getItems());
        };

        cbA.valueProperty().addListener((obs, oldV, newV) -> refreshAvailable.run());
        cbB.valueProperty().addListener((obs, oldV, newV) -> refreshAvailable.run());
        refreshAvailable.run();

        // Layout: Children area with two lists and add/remove buttons in between
        VBox buttonsBox = new VBox(8, btnAddChild, btnRemoveChild);
        buttonsBox.setAlignment(Pos.CENTER_LEFT);
        HBox childrenBox = new HBox(10, lvChildrenAvailable, buttonsBox, lvChildrenSelected);
        HBox.setHgrow(lvChildrenAvailable, Priority.ALWAYS);
        HBox.setHgrow(lvChildrenSelected, Priority.ALWAYS);
        lvChildrenAvailable.setPrefHeight(200);
        lvChildrenSelected.setPrefHeight(200);
        // Let the controls expand horizontally in column 1
        cbA.setMaxWidth(Double.MAX_VALUE);
        cbB.setMaxWidth(Double.MAX_VALUE);

        grid.add(lA, 0, 0); grid.add(cbA, 1, 0);
        grid.add(lB, 0, 1); grid.add(cbB, 1, 1);
        grid.add(lMarriageDate, 0, 2); grid.add(hbMarriageDate, 1, 2);
        grid.add(lMarriagePlace, 0, 3); grid.add(tfMarriagePlace, 1, 3);
        grid.add(lChildren, 0, 4); grid.add(childrenBox, 1, 4);

        return grid;
    }

    private static String nullToEmpty(String s) {
        return s == null ? "" : s;
    }

    private static String displayName(Individual i) {
        String f = i.getFirstName() != null ? i.getFirstName() : "";
        String l = i.getLastName() != null ? i.getLastName() : "";
        return (f + " " + l).trim();
    }
}

File: src/main/kotlin/com/family/tree/ui/FamilyDialog.kt
package com.family.tree.ui

import com.family.tree.model.Family
import com.family.tree.model.Gender
import com.family.tree.model.Individual
import com.family.tree.storage.ProjectRepository
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.layout.*
import javafx.stage.Modality
import javafx.stage.Stage
import java.util.*

class FamilyDialog {

    fun showCreate(data: ProjectRepository.ProjectData?): Optional<Family> {
        if (data == null) return Optional.empty()
        val stage = Stage()
        stage.initModality(Modality.APPLICATION_MODAL)
        stage.title = "New Family"

        val grid = buildForm(null, data)
        val root = VBox(grid).apply { padding = Insets(8.0) }
        val scene = Scene(root, 640.0, 520.0)
        stage.scene = scene
        stage.isResizable = true
        stage.minWidth = 580.0
        stage.minHeight = 460.0

        var result: Family? = null
        val btnOk = Button("Create")
        val btnCancel = Button("Cancel")
        val buttons = HBox(10.0, btnOk, btnCancel).apply { alignment = Pos.CENTER_RIGHT }
        GridPane.setColumnSpan(buttons, 2)
        grid.add(buttons, 0, 5)

        btnOk.isDefaultButton = true
        btnCancel.isCancelButton = true

        @Suppress("UNCHECKED_CAST")
        val cbA = grid.lookup("#husband") as ComboBox<Individual>
        @Suppress("UNCHECKED_CAST")
        val cbB = grid.lookup("#wife") as ComboBox<Individual>
        @Suppress("UNCHECKED_CAST")
        val lvChildrenSelected = grid.lookup("#childrenSelected") as ListView<Individual>

        btnOk.setOnAction {
            val fam = Family()
            val a = cbA.value
            val b = cbB.value
            fam.husbandId = a?.id
            fam.wifeId = b?.id
            lvChildrenSelected.items.forEach { child -> fam.childrenIds.add(child.id) }
            result = fam
            com.family.tree.util.DirtyFlag.setModified()
            stage.close()
        }
        btnCancel.setOnAction { stage.close() }

        stage.showAndWait()
        return Optional.ofNullable(result)
    }

    fun showEdit(fam: Family?, data: ProjectRepository.ProjectData?): Boolean {
        if (fam == null || data == null) return false
        val stage = Stage()
        stage.initModality(Modality.APPLICATION_MODAL)
        stage.title = "Edit Family"

        val grid = buildForm(fam, data)
        val root = VBox(grid).apply { padding = Insets(8.0) }
        val scene = Scene(root, 640.0, 480.0)
        stage.scene = scene
        stage.isResizable = true
        stage.minWidth = 580.0
        stage.minHeight = 420.0

        var saved = false
        val btnOk = Button("Save")
        val btnCancel = Button("Cancel")
        val buttons = HBox(10.0, btnOk, btnCancel).apply { alignment = Pos.CENTER_RIGHT }
        GridPane.setColumnSpan(buttons, 2)
        grid.add(buttons, 0, 3)

        btnOk.isDefaultButton = true
        btnCancel.isCancelButton = true

        @Suppress("UNCHECKED_CAST")
        val cbA = grid.lookup("#husband") as ComboBox<Individual>
        @Suppress("UNCHECKED_CAST")
        val cbB = grid.lookup("#wife") as ComboBox<Individual>
        @Suppress("UNCHECKED_CAST")
        val lvChildrenSelected = grid.lookup("#childrenSelected") as ListView<Individual>

        btnOk.setOnAction {
            val a = cbA.value
            val b = cbB.value
            fam.husbandId = a?.id
            fam.wifeId = b?.id
            fam.childrenIds.clear()
            lvChildrenSelected.items.forEach { child -> fam.childrenIds.add(child.id) }
            saved = true
            com.family.tree.util.DirtyFlag.setModified()
            stage.close()
        }
        btnCancel.setOnAction { stage.close() }

        stage.showAndWait()
        return saved
    }

    private fun buildForm(existing: Family?, data: ProjectRepository.ProjectData): GridPane {
        val grid = GridPane()
        grid.hgap = 10.0
        grid.vgap = 10.0
        grid.padding = Insets(12.0)
        val col0 = ColumnConstraints().apply {
            hgrow = Priority.NEVER
            isFillWidth = false
        }
        val col1 = ColumnConstraints().apply { hgrow = Priority.ALWAYS }
        grid.columnConstraints.addAll(col0, col1)

        val lA = Label("Husband:").apply { minWidth = Region.USE_PREF_SIZE }
        val cbA = ComboBox<Individual>().apply { id = "husband" }
        // Only males can be selected as husband
        data.individuals.forEach { i -> if (i.gender == Gender.MALE) cbA.items.add(i) }
        cbA.setCellFactory {
            object : ListCell<Individual>() {
                override fun updateItem(item: Individual?, empty: Boolean) {
                    super.updateItem(item, empty)
                    text = if (empty || item == null) null else displayName(item)
                }
            }
        }
        cbA.buttonCell = object : ListCell<Individual>() {
            override fun updateItem(item: Individual?, empty: Boolean) {
                super.updateItem(item, empty)
                text = if (empty || item == null) null else displayName(item)
            }
        }
        cbA.converter = object : javafx.util.StringConverter<Individual>() {
            override fun toString(obj: Individual?): String = obj?.let { displayName(it) } ?: ""
            override fun fromString(string: String?): Individual? = null
        }

        val lB = Label("Wife:").apply { minWidth = Region.USE_PREF_SIZE }
        val cbB = ComboBox<Individual>().apply { id = "wife" }
        // Only females can be selected as wife
        data.individuals.forEach { i -> if (i.gender == Gender.FEMALE) cbB.items.add(i) }
        cbB.setCellFactory(cbA.cellFactory)
        cbB.buttonCell = object : ListCell<Individual>() {
            override fun updateItem(item: Individual?, empty: Boolean) {
                super.updateItem(item, empty)
                text = if (empty || item == null) null else displayName(item)
            }
        }
        cbB.converter = object : javafx.util.StringConverter<Individual>() {
            override fun toString(obj: Individual?): String = obj?.let { displayName(it) } ?: ""
            override fun fromString(string: String?): Individual? = null
        }

        val lChildren = Label("Kids:").apply { minWidth = Region.USE_PREF_SIZE }
        val lvChildrenAvailable = ListView<Individual>().apply {
            id = "childrenAvailable"
            items.addAll(data.individuals)
            selectionModel.selectionMode = SelectionMode.MULTIPLE
            setCellFactory {
                object : ListCell<Individual>() {
                    override fun updateItem(item: Individual?, empty: Boolean) {
                        super.updateItem(item, empty)
                        text = if (empty || item == null) null else displayName(item)
                    }
                }
            }
        }
        val lvChildrenSelected = ListView<Individual>().apply {
            id = "childrenSelected"
            selectionModel.selectionMode = SelectionMode.MULTIPLE
            setCellFactory {
                object : ListCell<Individual>() {
                    override fun updateItem(item: Individual?, empty: Boolean) {
                        super.updateItem(item, empty)
                        text = if (empty || item == null) null else displayName(item)
                    }
                }
            }
        }

        val btnAddChild = Button("Add ▶")
        val btnRemoveChild = Button("◀ Remove")
        btnAddChild.setOnAction {
            val selected = ArrayList(lvChildrenAvailable.selectionModel.selectedItems)
            for (ind in selected) {
                if (!lvChildrenSelected.items.contains(ind)) {
                    lvChildrenSelected.items.add(ind)
                }
            }
            lvChildrenAvailable.selectionModel.clearSelection()
        }
        btnRemoveChild.setOnAction {
            val selected = ArrayList(lvChildrenSelected.selectionModel.selectedItems)
            lvChildrenSelected.items.removeAll(selected)
        }

        lvChildrenAvailable.setOnMouseClicked { e -> if (e.clickCount == 2) btnAddChild.fire() }
        lvChildrenSelected.setOnMouseClicked { e -> if (e.clickCount == 2) btnRemoveChild.fire() }

        if (existing != null) {
            existing.husbandId?.let { hid ->
                data.individuals.firstOrNull { it.id == hid }?.let { h ->
                    if (!cbA.items.contains(h)) cbA.items.add(0, h)
                    cbA.value = h
                }
            }
            existing.wifeId?.let { wid ->
                data.individuals.firstOrNull { it.id == wid }?.let { w ->
                    if (!cbB.items.contains(w)) cbB.items.add(0, w)
                    cbB.value = w
                }
            }
            for (cid in existing.childrenIds) {
                data.individuals.firstOrNull { it.id == cid }?.let { i ->
                    if (!lvChildrenSelected.items.contains(i)) lvChildrenSelected.items.add(i)
                }
            }
        }

        val refreshAvailable = {
            lvChildrenAvailable.items.setAll(data.individuals)
            val a = cbA.value
            val b = cbB.value
            if (a != null) lvChildrenAvailable.items.remove(a)
            if (b != null) lvChildrenAvailable.items.remove(b)
            if (a != null) lvChildrenSelected.items.remove(a)
            if (b != null) lvChildrenSelected.items.remove(b)
            lvChildrenAvailable.items.removeAll(lvChildrenSelected.items)
        }

        cbA.valueProperty().addListener { _, _, _ -> refreshAvailable() }
        cbB.valueProperty().addListener { _, _, _ -> refreshAvailable() }
        refreshAvailable()

        val buttonsBox = VBox(8.0, btnAddChild, btnRemoveChild).apply { alignment = Pos.CENTER_LEFT }
        val childrenBox = HBox(10.0, lvChildrenAvailable, buttonsBox, lvChildrenSelected)
        HBox.setHgrow(lvChildrenAvailable, Priority.ALWAYS)
        HBox.setHgrow(lvChildrenSelected, Priority.ALWAYS)
        lvChildrenAvailable.prefHeight = 200.0
        lvChildrenSelected.prefHeight = 200.0
        cbA.maxWidth = Double.MAX_VALUE
        cbB.maxWidth = Double.MAX_VALUE

        grid.add(lA, 0, 0); grid.add(cbA, 1, 0)
        grid.add(lB, 0, 1); grid.add(cbB, 1, 1)
        grid.add(lChildren, 0, 2); grid.add(childrenBox, 1, 2)

        return grid
    }

    private fun displayName(i: Individual): String {
        val fn = i.firstName ?: ""
        val ln = i.lastName ?: ""
        val name = "$fn $ln".trim()
        return if (name.isBlank()) i.id else name
    }
}
